### ISSUE TITLE
[Feature Request] GUI Text outline

### ISSUE THREAD

**Issue by makotech222** [xenko3d/xenko#102]
Allow adding a text outline to a UI Label. It is hard to distinguish text in game when you can't add an outline.

**Comment by xen2** [comment]
Agree, it could be quite useful.

**Comment by Kryptos-FR** [comment]
I wonder if that should be a feature of the UI control (here Label) or part of the font (if that's even possible).

**Comment by xen2** [comment]
@Kryptos-FR Good question! The problem is that ideally it should be on the Label, but that's only possible with dynamic fonts. Static fonts would need to have the outline backed in the precomputed bitmap.

**Comment by xen2** [comment]
Thinking more about it, it probably should do same as Italic/Bold since it's the same issue.
Seems like it's part of the font asset (in `FontProviderBase`).
We just have to make sure the same .ttf data & freetype memory is used if font is included twice.

**Comment by Kryptos-FR** [comment]
That would require the UI system to accept material, which it doesn't at the moment if I remember correctly.

**Comment by xen2** [comment]
Good point, it should be trivial with signed distance fields. Annoying this feature has different runtime/compile time requirements depending on type of font (runtime, static or signed distance field).

Maybe an option could be to first allow it at runtime only for SDF fonts (otherwise issue a warning) and not support it for other type of fonts.

Let me know what you think.

**Comment by mafiesto4** [comment]
Go with 1. (separate Epic task for the future?). For example here is a [sample tutorial](https://docs.flaxengine.com/manual/ui/tutorials/create-font-material.html) on how to create textured text in Flax. It could be possible to do sth similar in Xenko.

**Comment by xen2** [comment]
BTW, forget what I said earlier, forgot that italic/bold is usually a separate TTF.

I also prefer 1 (except it will just take a while).

**Comment by xen2** [comment]
It's definitely a good opportunity to leverage our material system better not only for text, but also for sprites and UI elements (something we wanted to do for a long time but never got around to do it).

**Comment by EssenRoc** [comment]
If can support text shadow, it’s even better.

**Comment by phr00t** [comment]
I found a good workaround here, that I made easier to use on my Focus fork: precompile a font to an image file (right click it in GameStudio). Open that image file and add outlines (or shadows). Standard Xenko packs the glyphs in really tight, so it doesn't give you much room to work with. I added an option to include a margin around letters to provide more room for effects:

https://github.com/phr00t/FocusEngine/commit/64c97dbced9a73c5130539bb6b0fea388439909a

**Comment by makotech222** [comment]
Forgot i even suggested this issue lol. I've been using Ultralight for my GUI now, so no longer an issue for me personally.

**Comment by Eideren** [comment]
Implemented through #2582 discussion around improvements should be done in #2584
